doc/manpages/ccnacl.1
'\" t
.\"     Title: ccnacl
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.75.2 <http://docbook.sf.net/>
.\"      Date: 01/10/2014
.\"    Manual: \ \&
.\"    Source: \ \& 0.8.2
.\"  Language: English
.\"
.TH "CCNACL" "1" "01/10/2014" "\ \& 0\&.8\&.2" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
ccnacl \- show and modify access control lists (ACLs) for access control of a CCNx content namespace\&.
.SH "SYNOPSIS"
.sp
\fBccnacl\fR [\-javaopts \fI<options>\fR] [\-debug \fI<portno>\fR] [\-as \fIkeystore\fR] \-show \fInodeName\fR
.sp
\fBccnacl\fR [\-javaopts \fI<options>\fR] [\-debug \fI<portno>\fR] [\-as \fIkeystore\fR] \-edit \fInodeName\fR \fIprincipalName\fR [null | r | rw | rw+]
.sp
\fBccnacl\fR [\-javaopts \fI<options>\fR] [\-debug \fI<portno>\fR] [\-as \fIkeystore\fR] \-init \fIdomain\fR \fIuserNamespace\fR \fIgroupNamespace\fR \fIprincipalName\fR
.SH "DESCRIPTION"
.sp
The \fBccnacl\fR utility prints or modifies an access control list (ACL) published with CCNx\&. An ACL specifies who should have access to a particular collection, identified by the given \fInodeName\fR\&. Entities that are supporting the CCNx access control profile will respect any ACL specification that applies to namespaces they are using\&. Read control is achieved through encryption, so as long as the writers support access control the readers must support it and be authorized in order to read content\&. Write control is achieved by consumers rejecting content that is not published by an authorized party\&.
.sp
The \fInodeName\fR specifies the CCNx namespace in which the ACL is published and to which it applies; CCNx Content Objects with \fInodeName\fR as a prefix of their names will be affected\&. The \fInodeName\fR must be specified using the CCNx URI encoding syntax\&. For simple cases of ASCII name components this is just pathname syntax with / delimiters\&.
.sp
An ACL is published as immutable CCNx content\&. When an ACL edit is performed, new versions of the ACL and related keys are published\&.
.SH "OPTIONS"
.PP
\fB\-as\fR \fIpathToKeystore\fR
.RS 4
Specifies a file path to a keystore; this is used for signing any newly\-created content\&. Applicable to all operations\&. The default keystore is used if this is not specified\&.
.RE
.PP
\fB\-show\fR \fInodeName\fR
.RS 4
Print the contents of the ACL for the given
\fInodeName\fR\&.
.RE
.PP
\fB\-edit\fR \fInodeName\fR \fIprincipalName\fR [null | r | rw | rw+]
.RS 4
Create an entry in the ACL of the given
\fInodeName\fR
to set the permissions for the given
\fIprincipalName\fR
as specified\&. If the ACL does not exist it is created\&. The possible permissions are:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}

\fBnull\fR
= no access
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}

\fBr\fR
= Reader
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}

\fBrw\fR
= Writer, can also read
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}

\fBrw+\fR
= Manager, can read, write, and edit ACLs
.RE
.RE
.PP
\fB\-init\fR \fIdomain\fR \fIuserNamespace\fR \fIgroupNamespace\fR \fIprincipalName\fR
.RS 4
Initialises the authentication domain\&.
.RE
.PP
\fB\-debug\fR \fIportno\fR
.RS 4
Allows attachment of an eclipse remote debugger to the utility at the specified port number\&.
.RE
.PP
\fB\-javaopts\fR \fIoptions\fR
.RS 4
Allows use of arbitrary java properties or options with the utility\&.
.RE
.SH "EXIT STATUS"
.PP
\fB0\fR
.RS 4
Success
.RE
.PP
\fB1\fR
.RS 4
Failure (syntax or usage error; retrieval error, failure to write)
.RE
.SH "FILES"
.sp
.if n \{\
.RS 4
.\}
.nf
$HOME/\&.ccnx/\&.ccnx_keystore \- default keystore
.fi
.if n \{\
.RE
.\}
.SH "AUTHOR"
.sp
Philippe Golle


NEWS
Changes between 0.8.1 and 0.8.2:
--------------------------------
This highlights some important changes since the most recent numbered release.
Refer to http://redmine.ccnx.org/versions/21 for more details on CCNx 0.8.2.

- This release features a refactored strategy layer, making it straightforward to add new
  strategies in a modular way.  Alongside the traditional "default" strategy there are
  two new examples, a "parallel" strategy that sends interests on all available faces
  and a "loadsharing" strategy that sends each arriving interest to just one of the
  available faces.  Strategies may be selected by prefix.  This selection may be
  done using new subcommands to the ccndc utility.

	* Feature #100929: Implement Junit test suites to facilitate bulk testing.
	* Feature #100983: strategy layer refactoring
	* Feature #100984: strategy labels on prefixes
	* Feature #100990: Restore SYNC_NOISY to INFO
	* Feature #100992: Add test-security to standard java tests
	* Feature #100993: Add test-versioning back into standard java tests
	* Feature #100994: Remove unused cruft from build.xml
	* Feature #101036: java sync should use Deque operations instead of Stack as recommended in the javadocs
	* Feature #101039: UDP mode for java netmanagers is not tested in standard testing
	* Feature #101053: ccncatchunks2 could use improved hole-filling and interest lifetime management
	* Bug #101030: build.xml setup of ccnr environment is incorrect
	* Bug #101033: Handling of extremely large values of MaxSuffixComponents is incorrect
	* Bug #101037: ccn_seqwriter library code should not limit the block size to 4096 bytes
	* Bug #101042: CcnFileProxy uses deprecated methods
	* Bug #101046: VLC 2.1.0 release necessitates change in plugin Makefile
	* Bug #101048: Do not rely on accept(2) returned addr for recognizing a unix-domain socket
	* Bug #101052: Unbreak objtree build target
	* Bug #101057: ccn_sign_content / ccn_chk_signing_params do not properly incorporate a timestamp in a template via ccn_signing_params
	* Bug #101058: ccndc in daemon mode does not exit cleanly when ccnd goes away
	* Bug #101059: testTimeouts a little too strict
	* Bug #101060: ccnd cache lookups should take a controllable amount of work
	* Bug #101061: ccnd minimum interest lifetime problem
	* Bug #101062: some ccnd logging missing check for logging enabled before calling expensive log procedure
	* Support #101047: Prepare for ccnx-0.8.2 release

Changes between 0.8.0 and 0.8.1:
--------------------------------
This highlights some important changes since the most recent numbered release.
Refer to http://redmine.ccnx.org/versions/20 for more details on CCNx 0.8.1.

- This release fixes a number of bugs, including a serious memory leak in ccnd.

	* Bug #101029: unbounded test log file growth
	* Bug #101032: ccnr should start at specified segment on checked start write
	* Bug #101034: No obvious documentation for how to start ccnr exists on website
	* Bug #101035: small leak in ccn_fetch should a realloc() fail
	* Bug #101038: Inconsistent encoding of zero-length name components
	* Bug #101043: ccnr r_store_lookup_backwards can fail
	* Bug #101044: valgrind reports memory leak in 0.8.0 ccnd
	* Bug #101049: ccnd can remove newly arrived content too quickly
	* Support #101045 Prepare for ccnx-0.8.1 release

Changes between 0.7.2 and 0.8.0:
--------------------------------
This highlights some important changes since the most recent numbered release.
Refer to http://redmine.ccnx.org/versions/17 for more details on CCNx 0.8.0.

- The use of symmetric key HMACs in place of public/private key signatures
  is now an option in both libraries.  In situations where it can be used,
  it provides a much higher performance alternative.
- The content store in ccnd has been improved.  It now uses a flatname
  representation for indexing, in the same way that ccnr does.  Also, the
  underlying skiplist implementation has moved into the library, so that
  applications may take advantage of it in the future.  In the process,
  some old ccnd bugs relating to the handling of extreme values of
  FreshnessSeconds were fixed.
- The more readable escaping convention in URIs is now the default.
  If for whatever reason you prefer to use only percent-escaping in URIs, use

	CFLAGS='-g -DCCN_URI_DEFAULT_ESCAPE=4' ./configure

  before typing make.  Note that both forms of escaping are accepted as input.
- An example rc startup script is now included, but needs to be installed by
  hand.  It is coded for Ubuntu Linux, but should be easily adaptable to other
  systems.
- This release drops support for Java JDK 1.5.

  * Issues Resolved:
	* Bug #100118: ccnd does not handle FreshnessSeconds=0 correctly.
	* Bug #100119: ccnd should handle large values of FreshnessSeconds
	* Bug #100558: test-security does not pass
	* Bug #100897: timeout values  should be consistently long throughout java code 
	* Bug #100966: Incorrect processing of CCNS_SYNC_SCOPE=3
	* Bug #100977: Problem with verify in CCNInputStream when trying to retrieve key
	* Bug #100987: Non-portable use of local keyword in Bourne shell scripts
	* Bug #100998: VLC 2.1.x block_New definition removed
	* Bug #100999: java BLOCK_AFTER_FIRST_SEGMENT doesn't work in all cases
	* Bug #101000: sync_diff code should reset differencing data when aborting or completing a comparison
	* Bug #101004: QueuedContentHandler has a sync error
	* Bug #101005: CCNx Android Services build out of sync with API changes to CCNx C-lib
	* Bug #101007: ccn_resolve_version() should distinguish between no response and finding an unversioned object.
	* Bug #101010: ccnseqwriter command busy-loops when its ccnd goes away
	* Bug #101011: Compile warning when building new symmetric key code with openssl 0.9.8
	* Bug #101012: Improve recovery from node fetch timeout in java sync
	* Bug #101013: node fetch timeout in java sync implementation should be longer and settable

javasrc/src/main/org/ccnx/ccn/utils/ccnacl.java
/*
 * A CCNx command line utility.
 *
 * Copyright (C) 2010, 2012 Palo Alto Research Center, Inc.
 *
 * This work is free software; you can redistribute it and/or modify it under
 * the terms of the GNU General Public License version 2 as published by the
 * Free Software Foundation.
 * This work is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
 * for more details. You should have received a copy of the GNU General Public
 * License along with this program; if not, write to the
 * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 * Boston, MA 02110-1301, USA.
 */

package org.ccnx.ccn.utils;

import java.io.File;
import java.util.ArrayList;
import java.util.logging.Level;

import org.ccnx.ccn.CCNHandle;
import org.ccnx.ccn.config.UserConfiguration;
import org.ccnx.ccn.impl.CCNFlowControl.SaveType;
import org.ccnx.ccn.impl.support.Log;
import org.ccnx.ccn.io.content.Link;
import org.ccnx.ccn.profiles.namespace.ParameterizedName;
import org.ccnx.ccn.profiles.security.access.AccessDeniedException;
import org.ccnx.ccn.profiles.security.access.group.ACL;
import org.ccnx.ccn.profiles.security.access.group.GroupAccessControlManager;
import org.ccnx.ccn.profiles.security.access.group.GroupAccessControlProfile;
import org.ccnx.ccn.profiles.security.access.group.ACL.ACLObject;
import org.ccnx.ccn.profiles.security.access.group.ACL.ACLOperation;
import org.ccnx.ccn.protocol.ContentName;

public class ccnacl {

	private static ContentName userStorage = new ContentName(UserConfiguration.defaultNamespace(), "Users");
	private static ContentName groupStorage = new ContentName(UserConfiguration.defaultNamespace(), "Groups");

	/**
	 * @param args
	 */
	public static void main(String[] args) {

		String extraUsage = "";

		// silence logging
		Log.setDefaultLevel(Level.WARNING);

		if ((args == null) || (args.length == 0)) {
			usage(extraUsage);
		}

		int pos = 0;
		if (args[0].startsWith("[")) {
			extraUsage = args[0];
			pos++;
		}
		if (args[pos].equals("-h")) {
			usage(extraUsage);
		}
		if (args[pos].equals("-as")) {
			if (args.length < pos+2) usage(extraUsage);
			pos++;
			setUser(args[pos]);
			pos++;
		}

		if (args[pos].equals("-show")) {
			if (args.length < pos + 2) {
				usage(extraUsage);
			}
			pos++;
			String nodeName = args[pos];
			showACL(nodeName);
		}
		else if (args[pos].equals("-edit")) {
			if (args.length < pos + 4) {
				usage(extraUsage);
			}
			String nodeName = args[pos + 1];
			String principalName = args[pos + 2];
			String role = args[pos + 3];
			if (! (role.equals("none") || role.equals(ACL.LABEL_READER) || role.equals(ACL.LABEL_WRITER) || role.equals(ACL.LABEL_MANAGER))) {
				usage(extraUsage);
			}
			editACL(nodeName, principalName, role);
		}
		else if (args[pos].equals("-init")) {
			if (args.length < pos + 5) {
				usage(extraUsage);
			}
			String domain = args[pos + 1];
			String userNamespace = args[pos + 2];
			String groupNamespace = args[pos + 3];
			String principalName = args[pos + 4];
			initACL(domain, userNamespace, groupNamespace, principalName);
		} else
			usage(extraUsage);
	}

	public static void usage(String extraUsage) {
		System.out.println("usage:");
		System.out.println("ccnacl " + extraUsage + "[-as pathToKeystore] -show nodeName");
		System.out.println("ccnacl " + extraUsage + "[-as pathToKeystore] -edit nodeName principalName [none|r|rw|rw+]");
		System.out.println("ccnacl " + extraUsage + "[-as pathToKeystore] -init domain userNamespace groupNamespace principalName");
		System.exit(1);
	}

	public static void setUser(String pathToKeystore) {
		File userDirectory = new File(pathToKeystore);
		String userConfigDir = userDirectory.getAbsolutePath();
		System.out.println("Loading keystore from: " + userConfigDir);
		UserConfiguration.setUserConfigurationDirectory(userConfigDir);
		// Assume here that the name of the file is the userName
		String userName = userDirectory.getName();
		if (userName != null) {
			System.out.println("User: " + userName);
			UserConfiguration.setUserName(userName);
		}
	}

	public static void showACL(String nodeName) {
		try{
			ContentName baseNode = ContentName.fromNative("/");
			GroupAccessControlManager acm = new GroupAccessControlManager(baseNode, groupStorage, userStorage, CCNHandle.open());
			ContentName node = ContentName.fromNative(nodeName);
			ACL acl = acm.getEffectiveACLObject(node).acl();
			System.out.println("ACL for node: " + nodeName);
			for (int j=0; j<acl.size(); j++) {
				Link lk = acl.get(j);
				System.out.println(lk.targetName() + " : " + lk.targetLabel());
			}
		}
		catch (Exception e) {
			e.printStackTrace();
			System.exit(1);
		}
		System.exit(0);
	}

	public static void editACL(String nodeName, String principalName, String role) {
		try{
			ContentName baseNode = ContentName.fromNative("/");
			GroupAccessControlManager acm = new GroupAccessControlManager(baseNode, groupStorage, userStorage, CCNHandle.open());
			ContentName node = ContentName.fromNative(nodeName);

			ACLObject initialACLObject = acm.getEffectiveACLObject(node);
			ACL initialACL = initialACLObject.acl();
			if (! initialACLObject.getBaseName().equals(GroupAccessControlProfile.aclName(node))) {
				// There is no actual ACL at this node.
				// So we copy the effective ACL to this node before updating it.
				acm.setACL(node, initialACL);
			}

			// initial role
			ContentName principal = ContentName.fromNative(principalName);
			Link plk = new Link(principal);
			String initialRole = null;
			for (int j=0; j<initialACL.size(); j++) {
				Link lk = initialACL.get(j);
				if (principal.compareTo(lk.targetName()) == 0) {
					initialRole = lk.targetLabel();
				}
			}

			// update
			ArrayList<ACLOperation> ACLUpdates = new ArrayList<ACLOperation>();
			if (initialRole == null) {
				if (role.equals(ACL.LABEL_READER)) ACLUpdates.add(ACLOperation.addReaderOperation(plk));
				else if (role.equals(ACL.LABEL_WRITER)) ACLUpdates.add(ACLOperation.addWriterOperation(plk));
				else if (role.equals(ACL.LABEL_MANAGER)) ACLUpdates.add(ACLOperation.addManagerOperation(plk));
			}
			else if (initialRole.equals(ACL.LABEL_READER)) {
				if (role.equals("none")) ACLUpdates.add(ACLOperation.removeReaderOperation(plk));
				else if (role.equals(ACL.LABEL_WRITER)) ACLUpdates.add(ACLOperation.addWriterOperation(plk));
				else if (role.equals(ACL.LABEL_MANAGER)) ACLUpdates.add(ACLOperation.addManagerOperation(plk));
			}
			else if (initialRole.equals(ACL.LABEL_WRITER)) {
				if (role.equals("none")) ACLUpdates.add(ACLOperation.removeWriterOperation(plk));
				else if (role.equals(ACL.LABEL_READER)) {
					ACLUpdates.add(ACLOperation.removeWriterOperation(plk));
					ACLUpdates.add(ACLOperation.addReaderOperation(plk));
				}
				else if (role.equals(ACL.LABEL_MANAGER)) ACLUpdates.add(ACLOperation.addManagerOperation(plk));
			}
			else if (initialRole.equals(ACL.LABEL_MANAGER)) {
				if (role.equals("none")) ACLUpdates.add(ACLOperation.removeManagerOperation(plk));
				else if (role.equals(ACL.LABEL_READER)) {
					ACLUpdates.add(ACLOperation.removeManagerOperation(plk));
					ACLUpdates.add(ACLOperation.addReaderOperation(plk));
				}
				else if (role.equals(ACL.LABEL_WRITER)) {
					ACLUpdates.add(ACLOperation.removeManagerOperation(plk));
					ACLUpdates.add(ACLOperation.addWriterOperation(plk));
				}
			}

			acm.updateACL(node, ACLUpdates);

			System.out.println("ACL for node: " + nodeName + " updated to assign role " + role + " to principal " + principalName);
		}
		catch (AccessDeniedException ade) {
			System.out.println("You do not have the permission to edit the acl at node: " + nodeName);
			System.exit(1);
		}
		catch (Exception e) {
			e.printStackTrace();
			System.exit(1);
		}
		System.exit(0);
	}

	// creates initial ACL at nodeName for principalName with permission rw+
	public static void initACL(String domain, String userNamespace, String groupNamespace, String principalName) {

		try{
			ContentName domainPrefix = ContentName.fromNative(domain);
			ContentName userNamespaceCN = ContentName.fromNative(userNamespace);
			ContentName groupNamespaceCN = ContentName.fromNative(groupNamespace);

			// Create the ACL for nodeName with principalName as a manager
			ArrayList<Link> ACLcontents = new ArrayList<Link>();
			Link lk = new Link(new ContentName(userNamespaceCN, principalName), ACL.LABEL_MANAGER, null);
			ACLcontents.add(lk);
			ACL domainRootACL = new ACL(ACLcontents);

			// Set user and group storage locations as parameterized names
			ArrayList<ParameterizedName> parameterizedNames = new ArrayList<ParameterizedName>();
			ParameterizedName uName = new ParameterizedName("User", userNamespaceCN, null);
			parameterizedNames.add(uName);
			ParameterizedName gName = new ParameterizedName("Group", groupNamespaceCN, null);
			parameterizedNames.add(gName);

			// Set access control policy marker
			ContentName profileName = ContentName.fromNative(GroupAccessControlManager.PROFILE_NAME_STRING);
			GroupAccessControlManager.create(domainPrefix, profileName, domainRootACL, parameterizedNames, null, SaveType.REPOSITORY, CCNHandle.open());

		} catch (Exception e) {
			e.printStackTrace();
			System.exit(1);
		}
		System.exit(0);
	}

}
